cassandra/cqlengine/concurrent.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2013-2017 DataStax, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from concurrent.futures import Future
+
+__all__ = ['CQLEngineFuture', 'CQLEngineFutureWaiter']
+
+
+class CQLEngineFuture(Future):
+    """
+    CQLEngineFuture provides a `concurrent.futures.Future` implementation to
+    work with the internal of CQLEngine.
+    """
+
+    _future = None
+    _post_processing = None
+    _result = None
+    _check_applied_fn = None
+
+    def __init__(self, future=None, post_processing=None, result=None):
+        super(CQLEngineFuture, self).__init__()
+        self._future = future
+
+        if self._future is None:
+            self.set_result(result)
+        else:
+            if post_processing:
+                self._post_processing = []
+                if isinstance(post_processing, (list, tuple)):
+                    self._post_processing += post_processing
+                else:
+                    self._post_processing.append(post_processing)
+
+            if isinstance(self._future, (CQLEngineFuture, CQLEngineFutureWaiter)):
+                self._future.add_done_callback(self._future_callback)
+            else:  # ResponseFuture
+                self._future.add_callbacks(self._response_future_callback, self._response_future_errback)
+
+    @classmethod
+    def _check_applied(cls, results):
+        """CQLEngine LWT check"""
+        if cls._check_applied_fn is None:
```

Comment:
We can't, since query.py imports things from concurrent.py. However, you make me realize that this is not for 3.x anymore ... so I could remove the check_applied function totally from there.

Reply:
This leads me to also move all Exceptions declaration from query.py to exceptions.py.

Reply:
Fixed. I decided to let all the check_applied logic to the query module.